man/man1/snap3dv.1
.TH SNAP3DV NEMO "8 September 1995"
.ds e =\h'-0.0m'
.SH NAME
snap3dv \- create ascii file for 3D viewing from a snapshot
.SH SYNOPSIS
\fBsnap3dv in=\fPsnapshot_file \fBout=\fPascii_file [parameters=values...]
.SH DESCRIPTION
\fIsnap3dv\fP prepares an ascii file in some "3d" format (see \fBMODE\fP below)
for viewing with
some of the available interactive 3D display tools that are available under
MSDOS, SunWindows and X-windows.
.PP
Three viewing coordinates must be selected, any general \fIbodytrans(5NEMO)\fP
variable can be choosen, though generally \fBx, y\fP and \fBz\fP will suffice.
A bounding box around the particles can be drawn,
a different cube can be selected, but particles outside the cube 
are not visible.
Each particle can be given its own
color, based on an integer \fIbodytrans\fP function (the actual
color on the screen wil depend on your device driver and hardware).
Lastly, any subset of particles can be choosen for display using a
visibility (0 or non-0) integer \fIbodytrans\fP function. For display tools
that support the multiple layer concept, the non-zero value of the
visibility is used to identify its layer. This often means that the
form of the \fBcolor\fP and \fBvisib\fP keywords are rather similar;
for example, 
.nf
.fi
.SH PARAMETERS
The following parameters are recognized in any order if the keyword is
also given:
.TP 25
\fBin=\fIin_file\fP
Input data file, must be in standard \fIsnapshot(5NEMO)\fP 
format [No default].
.TP
\fBxvar\*e\fP\fIx-expression\fP
The value of \fIx-expression\fP is plotted along the X axis.
\fIx-expression\fP is a real-valued C language expression over the
real values \fBx\fP, \fBy\fP, \fBz\fP, \fBvx\fP, \fBvy\fP, \fBvz\fP,
\fBphi\fP, \fBt\fP, and the integer value \fBi\fP.
The default is \fBx\fP.
.TP
\fByvar\*e\fP\fIy-expression\fP
See above.
The default is \fBy\fP.
.TP
\fBzvar\*e\fP\fIz-expression\fP
See above.
The default is \fBz\fP.
.TP
\fBxrange\*e\fP\fIx-axis-range\fP
Range of values for the X axis, specified either as a single number
(in which case the lower bound is zero) or a pair of numbers separated
by a colon or none, if the program is to use autoscaling.
The default range is autoscaling.
.TP
\fByrange\*e\fP\fIy-axis-range\fP
See above. The default is autoscaling. 
.TP
\fBzrange\*e\fP\fIz-axis-range\fP
See above. The default is autoscaling.
.TP
\fBtime\*e\fP\fItime\fP
Only the first plot frame with the correct time value 
will be selected for output.
The default is the first frame in a snapshot.
.TP
\fBvisib\*e\fP\fIexpression\fP
Determines the visibility of particles.
\fIexpression\fP is a integer expression in the above variables;
only points with non-zero values are plotted.
The default is \fB1\fP.
.TP
\fBcolor\*e\fP\fIexpression\fP
Determines the point color.
\fIexpression\fP is an integer expression in the above variables;
colors depend on your device driver/hardware.
The default is \fB10\fP.
.TP
\fBborder\*e\fP\fIborder-color\fP
Color of the border if to be plotted. set color to -1 if no border is
requested.
.TP
\fBformat\*e\fP\fIC-%f-fmt\fP
Floating point format with which the X,Y,Z coordinates are plotted.
Although the default format (\fB%f\fP) is general, one can generally
find a format which produced a more compact file format.
[Default: \fB%f\fP]
.TP
\fBmode\*e\fP\fIoutput_mode\fP
Output mode. Select any of the supported formats described in more
detail below: \fB3dv, x3d, acd, wld\fP or \fBdcx\fP. [Default: \fP3dv\fP].
.SH MODE: WLD
The \fIViewWld\fP is a Suntools program, and accepts a list of
vectors and/or points. The first 3 columns are the X,Y and Z
positions. The 4th column is an integer denoting move(0),
draw line(1) or draw point(2).
.SH MODE: 3DV 
The \fB3DV\fP format is a free format ascii listing consisting
of two parts. For each part the first line consists of an integer
denoting how many data lines this part consists of (excluding
that header line). This first part defines the X,Y,Z positions
of N points in 3D space:
.nf
         N
         X1 Y1 Z2
         X2 Y2 Z2
         ...
         XN YN ZN
.fi
The second part consists of the drawing operations, now referring
to the points by their index number (1 being the first one).
The point index number is in the first column; if the second column
is 0, it means a move (with no color), for a non-zero number it
is a draw to that next point with a color as specified by
by that integer. It will depend on your hardware what the color is.
Again, the first  line in this part consists of an integer how
many "line-segments" will follow.
.PP
For example, to draw N points withouth connecting lines, each point
needs 2 segments:
.nf
         i 0
         i color
.fi
.PP
NOTE: the \fI3dv\fP program can only handle 1500 or so stars. It then
quickly runs out of memory. (an old 1988 version of the program was
tested for this)
.SH MODE: X3D
The X3D format is used by a public domain program \fBx3d\fB.
It is a simple ascii file, with 3 sections. The first section is one line,
containing 2 integers: number of points, and number of segments. The next
two sections are the points and the segments.
Points have 3 real coordinates, segments two integers, describing
which two points are to be connected.
.nf
NP NS
X1 Y1 Z1
...
F1 T1
...
.fi
.SH MODE: ACD
\fIAcroSpin(tm)\fP format can be written in a very compact way:
5 columns with resp. X, Y, Z, Color and Layer, with a one line
header describing its format:
.nf
    Pointlist X Y Z Color Layer
    X1 Y1 Z1 C1 L1
    X2 Y2 Z2 C2 L2
    ...
.fi
Note that the total number of particles is not needed in this format.
.SH MODE: DCX
This (binary) format is used by programs such as \fIAutoCAD(tm)\fP. We do not
support it, but the MSDOS program ACROTRAN can translate such files into
\fIacd\fP format.
.SH "SEE ALSO"
bodytyrans(1NEMO), snapshot(5NEMO), snapxyz(1NEMO)
.nf
.ta +3i
ftp://dpls.dacc.wisc.edu/graphics/x3d		x3d.2.2.tar.Z (28-jun-94)
.fi
.SH AUTHOR
Peter Teuben
.SH FILES
.nf
.ta +1.5i
~/src/nbody/io/ 	snap3dv.c 
.fi
.SH "UPDATE HISTORY"
.nf
.ta +1.0i +4.0i
26-sep-91	V0.0: created for MSDOS 	PJT
9-oct-91	V1.0: adapted for NEMO to read snapshots instead	PJT
10-oct-91	V1.0a: added the ACD format	PJT
11-oct-91	V1.0b: added WLD format 	PJT
8-sep-95	V1.0c: added X3D format  	PJT
10-jan-96	V1.0d: upgraded for new X3D (V2.2) format	PJT
.fi


src/nbody/io/snap3dv.c
/*
 * SNAP3DV:  convert snapshot to 3dv format for 3D display
 *           (the output format is described in the manual page)
 *
 *      26-sep-91   Created             Peter Teuben
 *	 9-oct-91   full NEMO program with snapshots         PJT
 *	10-oct-91   option to write out AcroSpin(TM) format  PJT
 *	 3-feb-95   1.0c: simple xyzc format		     PJT
 *       7-sep-95   1.0d: added x3d format for point sources PJT
 *	20-sep-00   1.0e: added speck format for partiview   PJT
 */

#include <stdinc.h>
#include <getparam.h>
#include <vectmath.h>		/* otherwise NDIM undefined */
#include <filestruct.h>

#include <snapshot/snapshot.h>	
#include <snapshot/body.h>
#include <snapshot/get_snap.c>


string defv[] = {
    "in=???\n           Input snapshot",
    "out=???\n          Output ascii file for 3DV",
    "xvar=x\n           X-bodytrans variable to view",
    "yvar=y\n           Y-bodytrans variable to view",
    "zvar=z\n           Z-bodytrans variable to view",
    "xrange=\n          X-range of variables [autoscale]",
    "yrange=\n          Y-range of variables [autoscale]",
    "zrange=\n          V-range of variables [autoscale]",
    "time=\n		Time to select (default: first one)",
    "visib=1\n          Visibility bodytrans (0=no >=yes & layer number)",
    "color=3\n          Color of particles (bodytrans)",
    "border=0\n         Color and Color-1 of borders (0=not plotted)",
    "format=%f\n        Format used to store floating point coords X,Y,Z",
    "mode=3dv\n         Output mode {3dv, acd, dxf, wld, xyzc, x3d}",
    "VERSION=1.0c\n     3-feb-92 PJT",
    NULL,
};

string usage="convert snapshot for 3d viewing formats";

local void setrange(real *, string);

void nemo_main(void)
{
    stream instr, fo;
    char fmt[80];
    int i, n, color, nbody, bits, visib, nout, border, major_col, minor_col;
    real x,y,z, xmin,ymin,zmin, xmax,ymax,zmax, tsnap, tselect;
    real xrange[2], yrange[2], zrange[2];
    string xvar, yvar, zvar, realfmt, timestr, mode;
    rproc btrtrans(), xfunc, yfunc, zfunc;
    iproc btitrans(), cfunc, vfunc;
    Body *btab=NULL, *bp, *bq;

    /* Open files */
    instr = stropen(getparam("in"),"r");
    fo = stropen(getparam("out"),"w");

    /* get various bodytrans functions */
    xvar = getparam("xvar");       xfunc = btrtrans(xvar);
    yvar = getparam("yvar");       yfunc = btrtrans(yvar);
    zvar = getparam("zvar");       zfunc = btrtrans(zvar);
    cfunc = btitrans(getparam("color"));
    vfunc = btitrans(getparam("visib"));

    /* various remaining parameters */
    border = getiparam("border");
    realfmt = getparam("format");
    sprintf(fmt,"%s %s %s",realfmt,realfmt,realfmt);
    dprintf(0,"Format used in ascii 3dv output file=%s\n",realfmt);
    timestr = getparam("time");
    if (timestr!=NULL && *timestr!=NULL) tselect=getdparam("time");
    mode = getparam("mode");

    get_history(instr);
    for(n=0;;) {  /* read until first matched time found ; n = #stars found */
	get_history(instr);                         /* paranoia: */
        if (!get_tag_ok(instr, SnapShotTag))        /* if no more snapshots */
            break;                                  /* done with work */
        get_snap(instr, &btab, &nbody, &tsnap, &bits);  /* get snapshot */
        if(timestr!=NULL && *timestr!=NULL) {       /* if time select */
           if (tsnap != tselect) continue;		/* skip this time */
	}
        for (bp=btab, bq=btab, i=0; bp<btab+nbody; bp++, i++) {  /* select */
            x = xfunc(bp,tsnap,i);
            y = yfunc(bp,tsnap,i);
            z = zfunc(bp,tsnap,i);
            color = cfunc(bp,tsnap,i);      /* color of particle */
            visib = vfunc(bp,tsnap,i);      /* also called: the layer */
            if (visib>0) {                  /* visib can be any > 0 */
                if (n==0) {
                    xmin=xmax=x;
                    ymin=ymax=y;
                    zmin=zmax=z;
                }
                xmin=MIN(x,xmin); xmax=MAX(x,xmax);
                ymin=MIN(y,ymin); ymax=MAX(y,ymax);
                zmin=MIN(z,zmin); zmax=MAX(z,zmax);
                Pos(bq)[0] = x;
                Pos(bq)[1] = y;
                Pos(bq)[2] = z;
                Key(bq) = color;                   /* remember Key as color */
                Aux(bq) = visib;    /* remember Aux (a real) as layer/visib */
                if(bp!=bq)
                    bcopy(bp,bq,sizeof(Body));
                n++;
                bq++;
            } /* if(visib) */
        } /* for (bp) */
        dprintf(0,"Copied %d particles for 3dv\n",n);

        if (1) break;                               /* only do one for now */

    } /* for(;;) */
    strclose(instr);
    if (n==0) { 
	strdelete(fo,TRUE);
	error("Last read snapshot at time=%f with nbody=%d no match",
		tsnap,nbody);
    }

    dprintf(0,"Found min and max:\n  %s=%f %f\n  %s=%f %f\n  %s=%f %f\n",
            xvar,xmin,xmax,yvar,ymin,ymax,zvar,zmin,zmax);
    setrange(xrange,getparam("xrange"));
    setrange(yrange,getparam("yrange"));
    setrange(zrange,getparam("zrange"));
    if(xrange[0]!=xrange[1]) { xmin=xrange[0]; xmax=xrange[1]; }
    if(yrange[0]!=yrange[1]) { ymin=yrange[0]; ymax=yrange[1]; }
    if(zrange[0]!=zrange[1]) { zmin=zrange[0]; zmax=zrange[1]; }

    for(bp=btab, bq=btab, nout=0; bp<btab+n; bp++) {
        if(xmin>Pos(bp)[0] || Pos(bp)[0]>xmax ||
           ymin>Pos(bp)[1] || Pos(bp)[1]>ymax ||
           zmin>Pos(bp)[2] || Pos(bp)[2]>zmax)   continue;
        if(bp!=bq)
            bcopy(bp,bq,sizeof(Body));
        bq++;
        nout++;
    }
    if(n!=nout) printf("%d particles outside selected box\n",n-nout);
    if(nout==0) warning("No particles left to plot");
    if(nout>1500) warning("Too many particles for a good 3DV run");
    n=nout;

/** Now the different modes will be handled **/
  if (streq(mode,"3dv") || streq(mode,"3DV")) {             /** 3DV format **/


    if(border>0)				/* print header */
        fprintf(fo,"%d\n",n+8);
    else
        fprintf(fo,"%d\n",n);

    for (bp=btab, i=0; bp<btab+n; bp++, i++) {	/* all XYZ of stars */
        fprintf(fo,fmt,Pos(bp)[0],Pos(bp)[1],Pos(bp)[2]);  
        fprintf(fo,"\n");
    }

    if(border>0){				/* all XYZ of bounding box */
        fprintf(fo,fmt,xmin,ymin,zmin); fprintf(fo,"\n");
        fprintf(fo,fmt,xmin,ymin,zmax); fprintf(fo,"\n");
        fprintf(fo,fmt,xmin,ymax,zmin); fprintf(fo,"\n");
        fprintf(fo,fmt,xmax,ymin,zmin); fprintf(fo,"\n");
        fprintf(fo,fmt,xmax,ymax,zmax); fprintf(fo,"\n");
        fprintf(fo,fmt,xmax,ymax,zmin); fprintf(fo,"\n");
        fprintf(fo,fmt,xmax,ymin,zmax); fprintf(fo,"\n");
        fprintf(fo,fmt,xmin,ymax,zmax); fprintf(fo,"\n");
    }
    if (border>0)				/* header of segments */
        fprintf(fo,"%d\n",2*n+16);
    else
        fprintf(fo,"%d\n",2*n);   

    for (bp=btab, i=0 ; i<n; bp++, i++) {    	/* move and draw all stars */
        fprintf(fo,"%d 0\n%d %d\n",i+1,i+1,Key(bp));
    }
    if (border>0) {				/* draw the bounding box */
        major_col = border;             /* kludge */
        minor_col = border+1;
        fprintf(fo,"%d %d\n",n+2,0);
        fprintf(fo,"%d %d\n",n+1,major_col);
        fprintf(fo,"%d %d\n",n+3,major_col);
        fprintf(fo,"%d %d\n",n+1,0);
        fprintf(fo,"%d %d\n",n+4,major_col);
        fprintf(fo,"%d %d\n",n+6,minor_col);
        fprintf(fo,"%d %d\n",n+5,minor_col);
        fprintf(fo,"%d %d\n",n+7,minor_col);
        fprintf(fo,"%d %d\n",n+4,minor_col);
        fprintf(fo,"%d %d\n",n+7,0);
        fprintf(fo,"%d %d\n",n+2,minor_col);
        fprintf(fo,"%d %d\n",n+8,minor_col);
        fprintf(fo,"%d %d\n",n+3,minor_col);
        fprintf(fo,"%d %d\n",n+6,minor_col);
        fprintf(fo,"%d %d\n",n+8,0);
        fprintf(fo,"%d %d\n",n+5,minor_col);
    }
  } else if (streq(mode,"acd") || streq(mode,"ACD")) {        /** ACROSPIN **/
    fprintf(fo,"Pointlist X Y Z Color Layer\n");
    for (bp=btab; bp<btab+n; bp++) {	            /* all XYZ of stars */
        fprintf(fo,fmt,Pos(bp)[0],Pos(bp)[1],Pos(bp)[2]);
        fprintf(fo," %d %d\n",Key(bp),(int)Aux(bp));     /* color and layer */
    }
    strclose(fo);
  } else if (streq(mode,"xyzc") || streq(mode,"XYZC")) {          /** XYZC **/
    dprintf(0,"X Y Z Color\n");
    for (bp=btab; bp<btab+n; bp++) {	            /* all XYZ of stars */
        fprintf(fo,fmt,Pos(bp)[0],Pos(bp)[1],Pos(bp)[2]);
        fprintf(fo," %d\n",Key(bp));     		/* color */
    }
    strclose(fo);
  } else if (streq(mode,"x3d") || streq(mode,"X3D")) {            /** X3D **/
    dprintf(0,"NP NS\nX Y Z..\nP1 P2..\n");
    fprintf(fo,"%d %d\n",n,n);
    for (bp=btab; bp<btab+n; bp++) {	            /* all XYZ of stars */
        fprintf(fo,fmt,Pos(bp)[0],Pos(bp)[1],Pos(bp)[2]);
        fprintf(fo,"\n");
    }
    for (i=0; i<n; i++) {	                       /* segments to plot */
        fprintf(fo,"%d %d\n",i,i);
    }
    strclose(fo);
  } else if (streq(mode,"wld") || streq(mode,"WLD")) {         /** VIEWWLD **/
    for (bp=btab; bp<btab+n; bp++) {	            /* all XYZ of stars */
        fprintf(fo,fmt,Pos(bp)[0],Pos(bp)[1],Pos(bp)[2]);
        fprintf(fo," %d\n",2);     /* 2 = 'draw point' command for ViewWld */
    }
    strclose(fo);
  } else if (streq(mode,"dcx") || streq(mode,"DCX")) {         /** AutoCAD **/
    warning("DCX format not supported - use ACD and convert with ACROTRAN");
    strdelete(fo,TRUE);
  } else {
    error("Illegal output mode %s",mode);
    strdelete(fo,TRUE);
  }
}


local void setrange(real *rval, string rexp)
{
    char *cptr;

    if (rexp==NULL || *rexp==0) {
      rval[0] = rval[1] = 0.0;
    } else {
      cptr = strchr(rexp, ':');
      if (cptr != NULL) {
        rval[0] = atof(rexp);
	rval[1] = atof(cptr+1);
      } else {
        rval[0] = 0.0;
	rval[1] = atof(rexp);
      }
    }
}

